dig/graph.go

```diff
@@ -183,5 +183,38 @@ func (g *graph) registerConstructor(c interface{}) error {
 	}
 
 	g.nodes[objType] = &n
+
+	// object needs to be part of the graph to properly detect cycles
+	if cycleErr := g.detectCycles(&n); cycleErr != nil {
+		// if the cycle was detected delete from the graph
+		delete(g.nodes, objType)
+		return errors.Wrapf(cycleErr, "unable to register %v", objType)
+	}
+
+	return nil
+}
+
+// When a new constructor is being inserted, detect any present cycles
+func (g *graph) detectCycles(n *funcNode) error {
+	visited := make(map[string]bool)
+	return g.recursiveDetectCycles(n, visited)
+}
+
+// DFS and tracking if same node is visited twice
+func (g *graph) recursiveDetectCycles(n graphNode, visited map[string]bool) error {
```

Comment:
When used as a set and you don't really care about the value (just existence of element in map), empty structs are more efficient. Zero size. map[string]struct{}